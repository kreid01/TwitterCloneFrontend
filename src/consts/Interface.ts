export type IPost = {
    id:number,
    userName:string
    userAt:string
    userImg:string,
    postTextBody:string
    postMedia:string
    postDate:string
    commentCount:number
    likeCount:number
    retweetCount:number
    isLiked?:boolean
    isRetweeted?:boolean
    comments: Array<Comment>
}

export type Comment = {
    UserAt: string
    UserName: string
    UserImg: string 
    CommentBody: string
    CommentMedia?: string
}

export type INewComment = {
        commentBody: string;
        commentMedia: string;
        userAt: string;
        userName: string;
        userImg: string;
}