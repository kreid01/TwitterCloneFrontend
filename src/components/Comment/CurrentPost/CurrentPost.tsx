import React from "react";
import { Post } from "components/Post/Post";
import { CreateComment } from "../CreateComment/CreateComment";
import { FaArrowLeft } from "react-icons/fa";
import { IPost } from "../../../consts/Interface";
import { useGetComments } from "hooks/utils/useGetComments";
import { Comment } from "../Comment/Comment";
import {
  useIsCommenting,
  useUpdateIsCommenting,
} from "context/IsCommentingContext";
import { useGetUser } from "../../../context/UserContext";

interface Props {
  post: IPost;
  setPosts: React.Dispatch<React.SetStateAction<IPost[] | undefined>>;
  makeCurrentPost: (post: IPost) => void;
  posts: IPost[];
  isUsersPost: boolean;
}

export const CurrentPost: React.FC<Props> = ({
  setPosts,
  makeCurrentPost,
  isUsersPost,
  posts,
  post,
}) => {
  const { comments, setComments } = useGetComments(post.id);
  const postsComments = () => {
    if (typeof comments !== "undefined" && comments) {
      return comments.map((comment, i) => {
        return (
          <Comment
            comment={comment}
            index={i}
            setComments={setComments}
            comments={comments}
          />
        );
      });
    }
  };

  const user = useGetUser();
  const index = posts.indexOf(post);
  const toggleComment = useUpdateIsCommenting() as () => void;
  const isCommenting = useIsCommenting();

  return (
    <div className="mb-5 absolute bg-white h-full z-10 " data-testid="post">
      <header
        className="pl-4 pt-4 pb-1 w-full
            backdrop-blur-lg bg-slate-400 
            bg-opacity-5 flex"
      >
        <button
          data-testid="returnButton"
          onClick={toggleComment}
          className="mr-4 mb-4"
        >
          <FaArrowLeft />
        </button>
      </header>
      <div className="flex mb-4">
        <Post
          makeCurrentPost={makeCurrentPost as any}
          isUsersPost={isUsersPost}
          post={post}
          posts={posts}
          setPosts={setPosts}
        />
      </div>
      {isCommenting && (
        <CreateComment
          currentIndex={index}
          setPosts={setPosts}
          setComments={setComments}
          posts={posts}
          post={post}
        />
      )}
      {postsComments()}
    </div>
  );
};
