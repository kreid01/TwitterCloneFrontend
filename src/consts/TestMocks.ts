export const testPost = {
  id: 1,
  userName: "test-name",
  userAt: "testAt",
  userImg: "testImg",
  postTextBody: "testBody",
  postMedia: "test-media",
  postDate: "testDate",
  commentCount: 1,
  likeCount: 1,
  retweetCount: 1,
  isRetweeted: false,
  isLiked: false,
  retweetedBy: [1],
  comments: [],
};

export const testComment = {
  id: 1,
  userName: "test-name",
  userAt: "testAt",
  userImg: "testImg",
  commentBody: "testBody",
  commentMedia: "test-media",
  commentDate: "test-date",
};
export const testPosts = [testPost, testPost];

export const testUser = {
  userId: 1,
  userName: "test-name",
  userAt: "testAt",
  userImg: "testImg",
  userCoverImg: "testCoverImg",
  userEmail: "testEmail",
  postsIds: [1, 2],
  joinDate: "testDate",
  following: [1, 2],
  followers: [1, 2],
  email: "",
  password: "",
};
