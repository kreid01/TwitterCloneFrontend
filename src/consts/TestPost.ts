export const testPost = {
    id:1,
    userName:'testName',
    userAt:'testAt',
    userImg:'testImg',
    postTextBody:'testBody',
    postMedia:'testMedia',
    postDate:'testDate',
    commentCount:1,
    likeCount:1,
    retweetCount:1
}