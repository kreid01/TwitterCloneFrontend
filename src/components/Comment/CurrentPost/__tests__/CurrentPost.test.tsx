/* eslint-disable testing-library/prefer-screen-queries */
import React from "react";
import { CurrentPost } from "../CurrentPost";
import { testPost, testPosts } from "consts/TestMocks";

import { screen, render, cleanup, fireEvent } from "@testing-library/react";
import "@testing-library/jest-dom";
import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";

const mockFn = jest.fn;

afterEach(cleanup);
it("render without crashing", () => {
  render(
    <BrowserRouter>
      <CurrentPost
        isUsersPost={false}
        posts={testPosts}
        setPosts={mockFn}
        currentIndex={1}
        post={testPost}
        setToCurrentPost={mockFn}
        isCommenting={true}
        key={1}
        handleLike={mockFn}
        handleRetweet={mockFn}
        handleComment={mockFn}
        closeComment={mockFn}
      />
    </BrowserRouter>
  );
});

it("matched snapshot", () => {
  const tree = renderer
    .create(
      <BrowserRouter>
        <CurrentPost
          isUsersPost={false}
          posts={testPosts}
          setPosts={mockFn}
          currentIndex={1}
          post={testPost}
          setToCurrentPost={mockFn}
          isCommenting={true}
          key={1}
          handleLike={mockFn}
          handleRetweet={mockFn}
          handleComment={mockFn}
          closeComment={mockFn}
        />
      </BrowserRouter>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

it("return button when clicked functions", () => {
  render(
    <BrowserRouter>
      <CurrentPost
        isUsersPost={true}
        setPosts={mockFn}
        posts={testPosts}
        currentIndex={1}
        post={testPost}
        setToCurrentPost={mockFn}
        isCommenting={true}
        key={1}
        handleLike={mockFn}
        handleRetweet={mockFn}
        handleComment={mockFn}
        closeComment={mockFn}
      />
    </BrowserRouter>
  );
  const returnButton = screen.getByTestId("returnButton");
  fireEvent.click(returnButton);
  expect(mockFn).lastCalledWith();
});
