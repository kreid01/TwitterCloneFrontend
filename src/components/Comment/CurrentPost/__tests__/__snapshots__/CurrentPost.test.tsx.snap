// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matched snapshot 1`] = `
<div
  className="mb-5 absolute bg-white h-full z-10 "
  data-testid="post"
>
  <header
    className="pl-4 pt-4 pb-1 w-full
 backdrop-blur-lg bg-slate-400 
 bg-opacity-5 flex"
  >
    <button
      className="mr-4 mb-4"
      data-testid="returnButton"
      onClick={null}
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 448 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M257.5 445.1l-22.2 22.2c-9.4 9.4-24.6 9.4-33.9 0L7 273c-9.4-9.4-9.4-24.6 0-33.9L201.4 44.7c9.4-9.4 24.6-9.4 33.9 0l22.2 22.2c9.5 9.5 9.3 25-.4 34.3L136.6 216H424c13.3 0 24 10.7 24 24v32c0 13.3-10.7 24-24 24H136.6l120.5 114.8c9.8 9.3 10 24.8.4 34.3z"
        />
      </svg>
    </button>
  </header>
  <div
    className="flex mb-4"
  >
    <div
      className=" hover:bg-neutral-100 border-b-2 px-4 pt-4"
    >
      <div
        className="flex"
        data-testid="post"
        onClick={[Function]}
      >
        <img
          alt=""
          className="w-12 h-12 rounded-full"
          src="testImg"
        />
        <div
          className="flex-col pl-4"
        >
          <p
            className=""
          >
            <a
              href="/testAt"
              onClick={[Function]}
            >
              <strong>
                test-name
              </strong>
            </a>
            <span
              className="text-gray-500"
            >
              @
              testAt
              Â·
              testDat
            </span>
          </p>
          <p>
            testBody
          </p>
          <img
            alt=""
            className="w-80 pt-2 max-h-72 rounded-xl"
            data-testid="media"
            src="test-media"
          />
        </div>
      </div>
      <div
        className="ml-10"
      >
        <div
          className="flex"
        >
          <div
            className="comment-icon"
          >
            <svg
              data-testid="commentButton"
              fill="currentColor"
              height="1em"
              onClick={[Function]}
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": undefined,
                }
              }
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M256 32C114.6 32 0 125.1 0 240c0 49.6 21.4 95 57 130.7C44.5 421.1 2.7 466 2.2 466.5c-2.2 2.3-2.8 5.7-1.5 8.7S4.8 480 8 480c66.3 0 116-31.8 140.6-51.4 32.7 12.3 69 19.4 107.4 19.4 141.4 0 256-93.1 256-208S397.4 32 256 32z"
              />
            </svg>
          </div>
          <p
            className="mt-3.5 text-xs"
          >
            1
          </p>
          <div
            className="retweet-icon"
          >
            <svg
              data-testid="retweetButton"
              fill="currentColor"
              height="1em"
              onClick={[Function]}
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": "none",
                }
              }
              viewBox="0 0 640 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M629.657 343.598L528.971 444.284c-9.373 9.372-24.568 9.372-33.941 0L394.343 343.598c-9.373-9.373-9.373-24.569 0-33.941l10.823-10.823c9.562-9.562 25.133-9.34 34.419.492L480 342.118V160H292.451a24.005 24.005 0 0 1-16.971-7.029l-16-16C244.361 121.851 255.069 96 276.451 96H520c13.255 0 24 10.745 24 24v222.118l40.416-42.792c9.285-9.831 24.856-10.054 34.419-.492l10.823 10.823c9.372 9.372 9.372 24.569-.001 33.941zm-265.138 15.431A23.999 23.999 0 0 0 347.548 352H160V169.881l40.416 42.792c9.286 9.831 24.856 10.054 34.419.491l10.822-10.822c9.373-9.373 9.373-24.569 0-33.941L144.971 67.716c-9.373-9.373-24.569-9.373-33.941 0L10.343 168.402c-9.373 9.373-9.373 24.569 0 33.941l10.822 10.822c9.562 9.562 25.133 9.34 34.419-.491L96 169.881V392c0 13.255 10.745 24 24 24h243.549c21.382 0 32.09-25.851 16.971-40.971l-16.001-16z"
              />
            </svg>
          </div>
          <p
            className="mt-3.5 text-xs"
          >
            1
          </p>
          <div
            className="heart-icon"
          >
            <svg
              data-testid="likeButton"
              fill="currentColor"
              height="1em"
              onClick={[Function]}
              stroke="currentColor"
              strokeWidth="0"
              style={
                Object {
                  "color": "none",
                }
              }
              viewBox="0 0 512 512"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M462.3 62.6C407.5 15.9 326 24.3 275.7 76.2L256 96.5l-19.7-20.3C186.1 24.3 104.5 15.9 49.7 62.6c-62.8 53.6-66.1 149.8-9.9 207.9l193.5 199.8c12.5 12.9 32.8 12.9 45.3 0l193.5-199.8c56.3-58.1 53-154.3-9.8-207.9z"
              />
            </svg>
          </div>
          <p
            className="mt-3.5 text-xs"
          >
            1
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
